### PR TITLE
docs: add workload object labels ADR

### PR DESCRIPTION
**Pull request by kaynetu**
## Summary

- add ADR-079 defining declarative labels for generated workload objects
- make `JobSpec.workloadMetadata` the canonical label boundary across direct Job, Workflow, WorkloadRun, and Certification paths
- define immutable metadata and post-override queue consistency guarantees for Kueue, KAI Scheduler, Run:ai, and other label-based integrations

## Related Issue

Related to #212. The implementation will follow after the ADR is approved.

## Type of Change
- [ ] 🐛 Bug fix
- [ ] ✨ New feature
- [ ] 💥 Breaking change
- [x] 📚 Documentation
- [ ] 🔧 Refactoring
- [ ] 🔨 Build/CI

## Component(s) Affected
- [ ] API / CRDs
- [ ] Controller / Reconcilers
- [ ] Catalog / Workloads
- [ ] CLI (nvcrectl)
- [ ] Helm / Deployment
- [x] Documentation / CI
- [ ] Other: ____________

## Testing
- [x] Tests pass locally
- [ ] Manual testing completed
- [x] No breaking changes (or documented)

`make ci`

## Risk Assessment

Low. This PR changes documentation only and does not alter runtime behavior. ADR-079 remains Proposed; the API, controller, render, validation, and test changes will be submitted separately after approval.

## Checklist
- [x] Self-review completed
- [x] Commits are signed off for the DCO (`git commit -s`)
- [x] `make manifests generate` run (if `*_types.go` was modified) — no API files changed
- [ ] Golden files updated (if integration test output changed) — no integration output changed
- [x] Documentation updated (if needed)
- [x] Ready for review

### CODE REVIEW

File: docs/designs/079-workload-object-labels.md [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+- `api/v1alpha1/workflow_types.go`
+  - Add optional `WorkflowSpec.GangScheduler` using the existing shared type.
+    Protect its presence at the WorkflowSpec level and its complete value with
+    field-level equality, using the same two-rule pattern as workload metadata.
+- `api/v1alpha1/workloadrun_types.go`
+  - Add optional `WorkloadRunSpec.WorkloadMetadata`. The existing whole-spec
+    immutability rule covers it.
+- `api/v1alpha1/certification_types.go`
+  - Add optional `CategoryOptions.WorkloadMetadata`. The existing whole-spec
+    immutability rule covers global and per-category forms.
```
[the remark this answers is not in the document]
Comment (kaynetu) [reply docs/designs/079-workload-object-labels.md]: No API field is missing here, but the ADR did not explain why, so I clarified it in 39e030d rather than adding a field.

`CertificationSpec` embeds `CategoryOptions` with `json:",inline"` as its global defaults (`api/v1alpha1/certification_types.go`, the `CategoryOptions` embed at the end of the struct). That is how every existing global default already works. So the single `CategoryOptions.WorkloadMetadata` field is deserialized at **both** paths:

- global: `Certification.spec.workloadMetadata`
- per-category: `spec.categories[].options.workloadMetadata`

Adding a separate top-level `CertificationSpec.WorkloadMetadata` would actually be wrong: it would collide with the inlined JSON key.

Immutability and precedence were already covered but are now stated next to the field: the existing whole-spec `self == oldSelf` rule covers both forms (Decision 6), and `ResolveOptions` merges global then per-category per key with the per-category value winning (Decision 3). Decision 1 and the `certification_types.go` implementation bullet now say this explicitly, including a "do not add a second top-level field" note for the implementer.

**Review by asivanadi0 — COMMENTED** [review]
COMMENT review — not an approve / request-changes gate.

Read ADR-079 at HEAD `5e39783`, the PR body, issue #212, ADR-076/077, and the cited surfaces on `main` (`api/v1alpha1/{job,certification,workloadrun,workflow}_types.go`, `pkg/workload/{adapter,trainjob}.go`, `pkg/controller/{job,certification,workflow}_controller.go`, `pkg/platform/{gang_scheduler,image}.go`, `ResolveOptions`, `applyOverrides`). Problem statement and Option B vs Adapter-mutator rationale check out; Certification `CategoryOptions` `json:",inline"` for the global `spec.workloadMetadata` path is real (CodeRabbit's missing-field note is a false positive — the 5e39783 clarification is correct); OverrideSpec cannot touch a top-level `WorkflowSpec.gangScheduler` today (only `jobTemplate` / `dependencies` / `orchestration`). I am not ready to accept the record as written — concrete gaps below.

### What looks solid

- Rejecting `SetWorkloadLabels` on Adapter and keeping `WorkloadSpec` as a one-of union matches the type model (`MinProperties=1`/`MaxProperties=1`) and the existing post-`Build` label overlay in `job_controller.go` (~361–367).
- No transitive copy from `JobTemplate.metadata.labels` → TrainJob is the right default; labels are executable policy.
- Named `ApplyResolvedWorkflowTransforms` as the ADR-077 rule-of-three trigger, with workload-metadata last and gang before image, matches the disjoint-field / pinned-order story in ADR-077.
- Reserved-key reject (not silent overwrite) for `app.kubernetes.io/managed-by` and `nvcre.nvidia.com/*`, plus the presence + `self == oldSelf` pair for optional immutability, are the right CRD patterns given how JobSpec already pins sibling fields.
- Testing plan's override matrix (TrainJob label → B, runtime label → B, remove TrainJob restores A, remove runtime fails, bad `runtimeRef`) is the right shape for Decision 4.

### Concrete gaps before I would accept this ADR

### 1. README index skips ADR-078 (number collision with open #336)

This branch's `docs/designs/README.md` jumps `077 → 079`. Open PR #336 already owns **ADR-078** (`078-jobset-ownership.md`) and inserts that row. Merging #337 alone leaves a hole; merging both without coordination risks a duplicate-number fight or a silent drop. Please either (a) note in Notes/Implementation that **078 is reserved by #336** and keep the index gap intentional until that lands, or (b) rebase this branch after #336 so the index is `077 / 078 / 079` in one commit. Do not ship an index that pretends 078 does not exist while another Proposed ADR holds that number.

### 2. Workflow-level mutability of `jobTemplate.spec.workloadMetadata` vs “immutable scheduling intent”

Decision 6 makes `JobSpec.workloadMetadata` immutable once a Job exists, and Decision 4 makes `WorkflowSpec.gangScheduler` immutable. Workflow itself has **no** whole-spec `self == oldSelf` (unlike WorkloadRun/Certification). Overrides cannot rewrite `spec.gangScheduler` (OverrideSpec surface checks out), but a hand edit (or any future surface) can still change `spec.jobTemplate.spec.workloadMetadata` between orchestration groups/iterations. Consequences already admit override edits can affect *future* Jobs for the gang-scheduler queue; they do **not** say the same for arbitrary workload labels on a long-lived Workflow.

Please pick one and write it down:

- **Sticky Workflow contract:** pin immutability of `jobTemplate.spec.workloadMetadata` on Workflow (presence + equality, same two-rule pattern), or
- **Explicit divergence:** Consequences + Validation state that later groups may observe different TrainJob labels if the Workflow template is edited mid-flight, and that only `WorkflowSpec.gangScheduler` is the sticky cross-group invariant.

Silent “Jobs are immutable so we are fine” under-sells multi-group Workflows.

### 3. Asymmetric TrainJob restore vs runtime fail needs a one-line invariant

Decision 4 restores a removed TrainJob queue label on the generated Job, but fails (no silent repair) when a runtime Job/pod queue label is missing or conflicting. That asymmetry is defensible — NVCRE owns the post-`Build` TrainJob label write; runtime objects are dependency payloads the operator may have redirected — but Implementation/`ValidateResolvedGangScheduling` should state the invariant in one sentence so an implementer does not “helpfully” auto-restore runtime labels or, conversely, fail closed on a missing TrainJob key before the insert step. Please add that sentence next to the helper contract (and a negative test: runtime-restore must not happen).

### 4. `BuildObject` merge vs adapter-produced labels — conflict rule missing

Implementation says `BuildObject` “preserves any labels produced by an adapter and merges requested labels.” Today's `TrainJobAdapter.Build` sets no labels (only Name/Namespace), so this is latent. The moment an adapter stamps a key that also appears in `workloadMetadata.labels`, the ADR does not say whether user wins, adapter wins, or construction fails. Please pin a rule (I would want **fail on conflicting values, idempotent on same value**, consistent with Decision 4's insert/same/conflict), and cover it in the helper tests even if the current TrainJob adapter produces an empty map.

### 5. Admission CEL for reserved map keys is load-bearing and underspecified

Decision 6 claims OpenAPI/CEL on the shared `WorkloadMetadata` type rejects reserved keys with **no webhook**. This repo has no existing CEL-over-`map[string]string` key precedent. “Implementation will figure out the CEL” is how reserved-key holes ship. Please add a sketched rule (e.g. `self.labels.all(k, k != 'app.kubernetes.io/managed-by' && !k.startsWith('nvcre.nvidia.com/'))` plus the usual key/value syntax constraints) under Implementation, and keep the envtest admission cases as the proof — not Go-only validators. If a pure-CRD formulation cannot express the full Kubernetes label grammar, say what is enforced at admission vs defensively in Go.

### Smaller / clarifying

1. **Behavior change for existing `gangScheduler` users.** Consequences already note TrainJob queue labels appear when gang scheduling is configured. Please add an explicit Validation/Testing row: Certification/WorkloadRun with `gangScheduler` and **no** `workloadMetadata` still emits the resolved queue (including code default `default-queue`) on TrainJob metadata — so reviewers see the intentional delta for every current KAI/Run:ai cert, not only the new-field path.

2. **Soft structure.** ADR-078-style records carry an explicit Validation table; 079 only has a Testing plan. Not blocking if the five items above land, but a short Validation table (API admission / BuildObject / override matrix / Cert↔CLI parity / no pod-label leakage) would make acceptance greppable.

I am not approving on this pass — items 1–5 should be resolved (or explicitly deferred with named follow-ons) in the ADR text before acceptance. Happy to re-review a revision.

**Review by ndipebot — COMMENTED** [review]
Read this against the code on `main` rather than taking the citations on trust, and where the ADR makes a claim about what the API server will accept, I reproduced it on envtest instead of arguing it from the spec.

The good news first, so nobody re-litigates it. For a record this long the load-bearing claims hold up. `Build` really does set only name and namespace, and the Job controller really does overlay `app.kubernetes.io/managed-by` and `nvcre.nvidia.com/job` after it at `pkg/controller/job_controller.go:361-367`, so Decision 5 is describing something the code already does rather than inventing a seam. `WorkloadSpec` really is a `MinProperties=1` / `MaxProperties=1` union, which is what makes the argument against Option A work rather than just sound good. The `CategoryOptions` `json:",inline"` claim is right, so CodeRabbit's missing-field comment is a false positive and `39e030d` is the correct response to it. Whole-spec `self == oldSelf` is on both `CertificationSpec:455` and `WorkloadRunSpec:397` as claimed. `pkg/render/render.go:446` does build the workload independently of the controller, so the parity gap the shared builder closes is a real one. I checked the two upstream claims at source rather than from memory: Kueue reads `kueue.x-k8s.io/queue-name` off the TrainJob's top-level `metadata.labels` and does mutate `spec.suspend` by webhook, and the Kubeflow KAI guide does put `kai.scheduler/queue` on TrainJob metadata. `go test ./test/docspolicy/` passes, and every relative link and cited path resolves except the two files the ADR proposes to create.

Three things I would like fixed. Only the first is real work.

**1. Decision 6's CEL will not install.**

The schema enforces label-key syntax and the reserved keys through CEL over `Labels map[string]string` (line 70), and no bound is stated anywhere in the record. OpenAPI `pattern` cannot constrain map keys, so those rules have to be `self.all(k, ...)`, and the API server's static cost estimator will not accept that on an unbounded map.

I patched our own generated CRDs with exactly the rules Decision 6 describes and installed them into envtest at the pinned 1.37.0. CRD creation is refused:

```
estimated rule cost exceeds budget by factor of 1.297610x
x-kubernetes-validations estimated rule cost total for entire OpenAPIv3 schema exceeds budget
  (try ... adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared)
```

Certification is the tightest path, because the same type lands at `spec.workloadMetadata` and again inside `spec.categories[]` with `MaxItems=64`. Nested there, even the reserved-key rule, which has no regex in it at all, blows the per-expression budget by 20x. `maxProperties: 32` on the map plus a `maxLength` on the value was the only difference between the failing run and the passing one.

We already know about this trap and have written it down three times: `certification_types.go:78`, `certification_types.go:399`, and `workflow_types.go:730`. What makes it worth fixing in the record rather than leaving to implementation is that there is nothing here to copy from. Nothing in `api/v1alpha1/` iterates a map today, and the one label-key rule we do have (`queueLabelKey`, `workloadrun_types.go:337`) is only cheap because it is a scalar behind `MaxLength=317`. So please put the bounds in Decision 6 and the `job_types.go` bullet, say they are there for the cost budget and not as a product limit, and pick the number in the record. That cap is how many workload labels a user may set across four CRDs, which is an API decision rather than an implementation detail.

To keep it in proportion: the whole-schema overage measured 10.5x for me, not the 100x the error text suggests on a first read, and the envtest cases the testing plan already mandates would catch this before merge. It is an under-specified API limit, not something that reaches users.

**2.** References line 540 puts `JobTemplateSpec` in `job_types.go`. It lives at `workflow_types.go:70-78`, and `job_types.go` has no occurrence of it. The Implementation section already has this right, so the two sections disagree with each other.

**3.** Decision 6 works out that a field-level `self == oldSelf` does not fire when an optional field is added or removed, which is correct and is exactly why the parent-level `has()` rule is there. But `JobSpec` already carries three optional pointer fields with only the field-level rule: `nodeHealthMonitor:179`, `goodputMeasurement:217`, `bandwidthMeasurement:224`. All three say immutable, and all three can be added to or removed from a live Job today. The record derives the principle and then applies it only to its own field. One line saying whether we retrofit those or scope them out would stop the CRD shipping two meanings of "immutable".

I chased a dozen other things and dropped all of them, so nobody needs to reopen them. The ones that took longest to rule out: that Decision 4 bundles a second feature into a record scoped to #212, which I do not think holds, since it is motivated in Context and the consequence is stated plainly in Consequences; that `WorkflowSpec.gangScheduler` is assertion-only; that Decision 7 contradicts Decision 4 about ADR-076; that the `ResolveOptions` no-aliasing instruction is unsatisfiable as pointed at, which it is not, because it targets `resolved := *global` at `certification_controller.go:623` and that precedes the early return; and that the new unsupported-runtime-shape rule contradicts ADR-076's deliberately tolerant helper. None of them survived contact with the code.

@asivanadi0 items 1, 3, 4 and 5 look right to me, and 5 is the same concern as my item 1 with a reproduction attached now. Item 2 I think is already answered by the type system, so worth settling before @kaynetu writes anything against it. `JobTemplateSpec.Spec` is a plain `JobSpec` at `workflow_types.go:78`, so markers on `JobSpec` fields land in the Workflow CRD too, and you can see the propagation working: the existing `workload is immutable` rule is in both `nvcre.nvidia.com_jobs.yaml` and `nvcre.nvidia.com_workflows.yaml`. So lines 234-243 are accurate as written, and a hand edit to `jobTemplate.spec.workloadMetadata` mid-flight would already be rejected.

Numbering is fine. #336 holds 078 and #338 holds 080, so the README gap is merge ordering and not a collision.

The part of this I would keep regardless of the above is the Rationale table and the "labels are executable policy" argument. That is what will stop someone six months from now wiring `JobTemplate.metadata.labels` straight through to the TrainJob because it looked convenient.

**Review by ndipebot — COMMENTED** [review]
Re-reviewed at `bd72853`. Everything I raised and everything @asivanadi0 raised is addressed. Rather than check the text against my earlier objection, I rebuilt the envtest reproduction against the bounds `86367df` actually prescribes, and the fix works. Two corrections below, both to text the response commit introduced, and neither blocks merge.

### My item 1 (Decision 6's CEL will not install): fixed, and I verified it

I rebuilt the patch from what Decision 6 and the `job_types.go` bullet now say, at all five sites the record names: `spec.workloadMetadata` on jobs, `spec.jobTemplate.spec.workloadMetadata` on workflows, `spec.workloadMetadata` on workloadruns, and both `spec.workloadMetadata` and `spec.categories[].options.workloadMetadata` on certifications. Installed into envtest at the pinned 1.37.0.

Without the bounds it still fails the way it did before:

```
properties[spec].properties[workloadMetadata].properties[labels].x-kubernetes-validations[1].rule:
  Forbidden: estimated rule cost exceeds budget by factor of 1.258288x
properties[categories].items.properties[options].properties[workloadMetadata].properties[labels].x-kubernetes-validations[0].rule:
  Forbidden: estimated rule cost exceeds budget by factor of 20.1x
spec.validation.openAPIV3Schema:
  Forbidden: x-kubernetes-validations estimated rule cost total for entire OpenAPIv3 schema
  exceeds budget by factor of 14.6x
```

With the bounds the record now prescribes, all four CRDs install. The key-grammar rules `86367df` added on top of what I originally tested do not blow the budget at 32.

I also checked the schema is functional and not merely installable. Against it the API server rejects `app.kubernetes.io/managed-by` and `nvcre.nvidia.com/job` with the reserved-key message, rejects `bad key` and `a/b/c` as malformed, accepts a 63-character key name and rejects 64, accepts a 253-character prefix and rejects 254, accepts an empty value and a 63-character value and rejects 64, accepts 32 labels and rejects 33, and rejects all three update transitions: value change, removal, and addition. That is the whole contract Decision 6 claims, enforced at admission with no webhook, which also settles @asivanadi0's item 5.

One detail worth recording for whoever implements this: the field-level `self == oldSelf` belongs on `JobSpec` only, as Decision 6 already says. I tried it on the nested `spec.categories[].options.workloadMetadata` path as well and the API server refuses the CRD outright with `oldSelf cannot be used on the uncorrelatable portion of the schema within spec.categories`. The record's choice to lean on the existing whole-spec rule for Certification and WorkloadRun is not just economical, it is the only thing that installs.

### Correction 1: Decision 6's reasoning about the bounds

Lines 243 to 246 say the bounds "are chosen to keep CEL map iteration within the API server's static cost budget" and that "the 32-entry cap is motivated by that budget". Neither half survives measurement:

1. Only `maxProperties` is load-bearing. `maxProperties: 32` with no value `maxLength` installs. A value `maxLength: 63` with no `maxProperties` is still rejected at `1.258288x`. The 63 is simply the Kubernetes label-value limit, which line 242 already states on its own.
2. The budget does not force 32. With the full prescribed rule set I get ACCEPTED at 32, 64, 128, 1024, and 4096. The first failure is 8192 at `1.7x`, then 16384 at `3.4x`. The ceiling sits roughly 128x above the number in the record.

So 32 is a product decision, which is the thing I asked to be made explicit rather than left to implementation. The record already says the better version of this one sentence later, "is an enforced API limit and must be documented as such". I would drop the causal clause and keep that, so the record does not carry two different justifications for the same number.

### Correction 2: the numbering Note is now false

Lines 586 to 589 say ADR-078 "is held by" open PR #336, and that "the design index on this branch therefore runs `077` to `079` on purpose". `bd72853` merged `main`, so #336 has landed and `docs/designs/README.md` on this branch already reads `077 / 078 / 079` with no gap. The `#338` and ADR-080 half is still accurate. This was true when it was written and the merge invalidated it.

### Nit on the type sketch

Line 70 declares `Labels map[string]string`, but the bullet at line 306 asks for `additionalProperties.maxLength: 63` and a value `pattern`. `controller-gen` v0.20.0 has no map analogue to its array-only `items:` markers, so neither marker can be written on that field. A named string value type carrying the markers generates the schema you want. Worth a word in the sketch so the implementer does not have to discover it.

### Loose end from my item 3

The scope-out line in Decision 6 and the follow-on bullet in Notes are what I asked for. Notes says the retrofit is "to be filed as its own issue before this record is accepted", and that issue does not exist yet. #212 is the only open one in this area.

### The rest

My item 2 is fixed: References now splits `job_types.go` and `workflow_types.go`, and `JobTemplateSpec` is at `workflow_types.go:71` with no occurrence in `job_types.go`.

On @asivanadi0's item 2, I checked the rebuttal I made rather than leaving it as an assertion. `workload is immutable` appears in both `nvcre.nvidia.com_jobs.yaml` and `nvcre.nvidia.com_workflows.yaml`, so `JobSpec` markers do propagate through `JobTemplateSpec` into the Workflow CRD, and a mid-flight hand edit to `jobTemplate.spec.workloadMetadata` is already rejected.

@asivanadi0's items 3 and 4 landed as asked, including the negative test that a missing runtime label stays absent after failed validation, and the insert/same/conflict rule for `BuildObject`. The Validation table and the explicit rows for existing `gangScheduler` users with an omitted queue cover the two smaller points.

**Comment by kaynetu** [comment]
Thanks for the review. Addressed the last few nits. @ndipebot

**Review by ndipebot — APPROVED** [review]

**Review by asivanadi0 — APPROVED** [review]
APPROVE — re-review at HEAD `eb25c6bb6f49024e5c1ebc7170aed7e4561445f0` (tip after `86367df` / `bd72853` / re-review commit).

Prior COMMENT was on `5e39783`. Re-read ADR-079 + README index at this tip and checked each previously flagged gap against the text (not against reply threads).

### Closed since `5e39783`

1. **README index / ADR-078** — Index now reads `077 / 078 / 079 / 080` (078 landed via #336 merge into this branch). Hole closed.
2. **Workflow `jobTemplate.spec.workloadMetadata` mutability** — Sticky contract chosen and written: Decision 6's presence + equality pair covers typed Workflow Job templates; Validation table row makes acceptance greppable; Consequences spell that typed metadata cannot be added/removed/changed after creation (overrides remain a separate surface, with immutable `WorkflowSpec.gangScheduler` as the sticky cross-group queue invariant).
3. **TrainJob restore vs runtime fail invariant** — One-line contract is in Implementation next to `ValidateResolvedGangScheduling` (restore missing TrainJob key; reject missing/conflicting runtime keys without repair). Testing plan asserts restore-A / fail-runtime / "missing runtime label remains absent after failed validation".
4. **`BuildObject` adapter-label conflict rule** — Decision 5 + Implementation pin fail-on-conflict / idempotent-on-same / unrelated keys survive. Helper tests cover the empty-adapter case.
5. **Reserved-key CEL sketch** — Implementation now sketches `self.all(k, k != 'app.kubernetes.io/managed-by' && !k.startsWith('nvcre.nvidia.com/'))` plus key-grammar bounds; admission must enforce the full contract; envtest CRD install is the proof.

### Smaller items also landed

- Existing `gangScheduler` / no-`workloadMetadata` behavior called out in Testing plan + Validation ("Existing users" row) — intentional TrainJob queue-label delta is visible.
- Validation table added (admission / template immutability / builder / gang / existing users / CLI parity / no pod-label leakage).

No residual gaps from the prior pass. Approving the record at this tip.
